server/build/webpack/base.config.js

```diff
@@ -0,0 +1,318 @@
+    }))
+  }
+
+  const dirNodeModules = path.join(dir, 'node_modules')
+  if (fs.existsSync(dirNodeModules)) {
+    nodeExternals({
+      modulesDir: dirNodeModules,
+      includeAbsolutePaths: true,
+      whitelist: [/es6-promise|\.(?!(?:js|json)$).{1,5}$/i]
+    })
+  }
+
+  return externals
+}
+
+export default async function baseConfig (dir, {dev = false, isServer = false, buildId, config}) {
+  const extractCSS = new ExtractTextPlugin({
+    filename: 'static/style.css',
+    disable: dev
+  })
+
+  const cssLoader = {
+    loader: isServer ? 'css-loader/locals' : 'css-loader',
+    options: {
+      modules: false,
+      minimize: !dev,
+      sourceMap: dev,
+      importLoaders: 1,
+      ...(config.cssLoaderConfig || {})
+    }
+  }
+
+  const postcssLoader = {
+    loader: 'postcss-loader',
+    options: {
+      plugins: () => {}
+    }
+  }
+
+  function cssLoaderConfig (loader = false) {
+    return [
+      isServer && !cssLoader.options.modules && 'ignore-loader',
+      isServer && cssLoader.options.modules && cssLoader,
+      isServer && cssLoader.options.modules && postcssLoader,
+      isServer && cssLoader.options.modules && loader,
+      ...(!isServer ? extractCSS.extract({
+        use: [cssLoader, postcssLoader, loader].filter(Boolean),
+        // Use style-loader in development
+        fallback: {
+          loader: 'style-loader',
+          options: {
+            sourceMap: true,
+            importLoaders: 1
+          }
+        }
+      }) : [])
+    ].filter(Boolean)
+  }
+
+  const babelLoaderOptions = babelConfig(dir, isServer)
+
+  let webpackConfig = {
+    devtool: dev ? 'cheap-module-source-map' : false,
+    // devtool: 'source-map',
+    name: isServer ? 'server' : 'client',
+    cache: true,
+    profile: true,
+    target: isServer ? 'node' : 'web',
+    externals: externalsConfig(dir, isServer),
+    context: dir,
+    entry: async () => {
+      const pages = await getPages(dir, {dev, isServer})
+      const mainJS = require.resolve(`../../../client/next${dev ? '-dev' : ''}`)
+      const clientConfig = !isServer ? {
+        'main.js': [
+          dev && !isServer && path.join(__dirname, '..', '..', '..', 'client', 'webpack-hot-middleware-client'),
+          dev && !isServer && path.join(__dirname, '..', '..', '..', 'client', 'on-demand-entries-client'),
+          mainJS
+        ].filter(Boolean)
+      } : {}
+      return {
+        ...clientConfig,
+        ...pages
+      }
+    },
+    output: {
+      path: path.join(dir, '.next', isServer ? 'dist' : ''),
+      filename: '[name]',
+      libraryTarget: 'commonjs2',
+      publicPath: `/_next/webpack/`,
+      // This saves chunks with the name given via require.ensure()
+      chunkFilename: '[name]-[chunkhash].js',
+      sourceMapFilename: '[file].map?[contenthash]'
+    },
+    performance: { hints: false },
+    resolve: {
+      extensions: ['.js', '.jsx', '.json', '.ts', '.tsx'],
```

Comment:
Let's do not ship `.ts` `.tsx` by default. Let the user to define it via the custom webpack config.